### PR TITLE
GORA-419: AccumuloStore.put deletes entire row when updating map/array field

### PR DESCRIPTION
**Pull request by gerhardgossen**
JIRA issue: https://issues.apache.org/jira/browse/GORA-419

### CODE REVIEW

File: gora-accumulo/src/main/java/org/apache/gora/accumulo/store/AccumuloStore.java [hunk 1]
```diff
@@ -746,12 +747,12 @@ private int putMap(Mutation m, int count, Schema valueType, Object o, Pair<Text,
     return count;
   }
 
-  private int putArray(Mutation m, int count, Object o, Pair<Text, Text> col) {
+    private int putArray(Mutation m, int count, Object o, Pair<Text, Text> col, String fieldName) {
```
Comment (hsaputra) [inline gora-accumulo/src/main/java/org/apache/gora/accumulo/store/AccumuloStore.java]: Accidental misaligned on the patch?

**Comment by renato2099** [comment]
Besides Henry's comment, I think this looks to me, thanks for the patch Gerhard!
Can you merge this @hsaputra please? or do you want me to do it?

File: gora-accumulo/src/main/java/org/apache/gora/accumulo/store/AccumuloStore.java (hunk 1 above)
Comment (gerhardgossen) [inline gora-accumulo/src/main/java/org/apache/gora/accumulo/store/AccumuloStore.java]: @hsaputra correct, I pushed a new commit with proper indentation.

File: gora-accumulo/src/main/java/org/apache/gora/accumulo/store/AccumuloStore.java [hunk 2]
```diff
@@ -718,12 +718,13 @@ else if (effectiveSchema.getType() == Type.MAP) {
     }
   }
 
-  private int putMap(Mutation m, int count, Schema valueType, Object o, Pair<Text, Text> col) throws GoraException {
+    private int putMap(Mutation m, int count, Schema valueType, Object o, Pair<Text, Text> col,
```
Comment (hsaputra) [inline gora-accumulo/src/main/java/org/apache/gora/accumulo/store/AccumuloStore.java]: With the update, the Pair for col only used to get the name, could we just use the fieldName instead for the Query?
I.E remove the col input param in putMap and putArray.

**Comment by renato2099** [comment]
Hi @gerhardgossen!
After taking a closer look, I have some questions, I might be totally off, but please bear with me.
The col pair that @hsaputra is referring to contains the family name and the qualifier used. So when putting the values, shouldn't we use something like ...setFieldName(col.getFirst() + col.getSecond()) to actually use the qualifier that was specified on the mapping?
And then what we would need to change is the deleteByQuery, that depending if it is a map, then we'd need to again the col pair and delete using it?

**Comment by gerhardgossen** [comment]
@renato2099: you need to distinguish between the *field name* (e.g. 'content') that is used in the external API (for example as members of the value objects) and the *column name* (e.g. 'f:cnt') that is used in the data store (in this case Accumulo). The mapping between both is specified in the `gora-*-mapping.xml` files.

`deleteByQuery` is part of the external API and therefore receives field name values (as part of the query parameter). If we change this method it would break user code that tries to delete a map/array field. As a map is represented in Accumulo as a collection of fields, we would also need to delete `cf:*` instead of a single `cf:qualifier`.

File: gora-accumulo/src/main/java/org/apache/gora/accumulo/store/AccumuloStore.java (hunk 2 above)
Comment (gerhardgossen) [inline gora-accumulo/src/main/java/org/apache/gora/accumulo/store/AccumuloStore.java]: We still need the column (at least the column family) further down to store the map/array values, so we need to keep the parameter.

**Comment by gerhardgossen** [comment]
This actually seems to be a bigger problem: As no mapping for the fields is found, the query passed to `deleteByQuery` will have no fields set. This means that **the entire row is removed even though we just wanted to remove a single CF** :boom:. I have created a [test case](https://github.com/gerhardgossen/gora-419) to demonstrate this. The test case is set up for Gora 0.5, because I couldn't get Nutch working on 0.6 yet. This code in AccumuloStore did not change between the releases, so it should still apply. Can anyone else try to reproduce it 0.6?

Is it possible to create a bugfix relase including this patch for 0.5, because Nutch does not support 0.6 yet (see NUTCH-1946).

**Comment by hsaputra** [comment]
@gerhardgossen Yes, you could file issue against 0.5.x so we could create new patch release for 0.5.x release.

**Comment by renato2099** [comment]
Thanks a lot for the explanation @gerhardgossen! And yes this is a problem we have seen in other data stores as well. I mean managing complex data types because not all data stores provide the same functionality. For example, in gora-cassandra depending on your mapping file, you could create subcolumns inside a super column or even separated columns. Then when updating maps, you could end up updating a whole column even when a single value was modified inside an array or map. This behaviour is of course wrong. I guess this is also happening in accumulo per your test.
I think there is a trade-off here between generating a column for each specific value of a map/array which leads to a more complex scan operation or using a single column to store them all which leads to the current behaviour.
In Cassandra, arrays and maps can be now stored natively, so I guess we will be using them soon instead of adding this extra "mapping" complexity. Do you know if Accumulo stores complex data types or if it plans to?

**Comment by gerhardgossen** [comment]
@renato2099: Unfortunately I don't know enough about Accumulo to know if there may be a better way. But I think this issue needs to be fixed because it causes data loss in unrelated fields.

Depending on the use case it might also make sense to disable this clear-before-update behavior, maybe even on a field-by-field basis. For example in Nutch 2, inlinks to a web page are store as a map (key: URL, value: anchor text). In this scenario it would be nice to be able to just add one more KV pair to the map without having to fetch, update and store the entire map.

**Comment by renato2099** [comment]
@gerhardgossen so you mean  just to overwrite the value without checking if it exists or not? I think this is also something that makes sense, but this would still mean that you would be writing the whole map even when a single value has been added. Thus, this means you would have to get the old map, add the new kv pair and write the updated map. This is not so efficient but we wouldn't loose previous values. Would you agree with this get-update-store behaviour?

**Comment by gerhardgossen** [comment]
@renato2099: do you mean the clear-before-put behavior? You probably don't need to fetch the old map if you're just adding new KVs, because the new KVs will just add new columns to the map's column family. So if you just update with the new values, new KVs will be added (as expected), new values for existing keys will overwrite the old values (as expected) and not mentioned keys will keep their existing value (which may or may not be what you expect (but e.g. Nutch seems to expect this)). For this you don't need any read or delete operations.

Anyway, this is probably outside of the scope of this pull request. The original bug is that if I update a map or array field, the values of completely unrelated columns will be deleted from the store (see [test case](https://github.com/gerhardgossen/gora-419)), which is always unexpected ;).

**Comment by lewismc** [comment]
Folks are we good to merge this PR? @renato2099 do you have anything to add?

**Comment by renato2099** [comment]
hey guys, let's just push it forward, and deal with it later on. IMO the changes proposed by  @gerhardgossen make sense and I am happy he took the time to nail it down :) Thanks @gerhardgossen !

**Comment by lewismc** [comment]
This has been committed folks. I don't know how the hell to close off these issues anymore though! It seems to not be happening automatically anymore. Any ideas?

**Comment by gerhardgossen** [comment]
Thanks for merging @lewismc, I'll close it from my side.

**Comment by hsaputra** [comment]
@lewismc Need to add "This closes #23" in the commit comment

**Comment by lewismc** [comment]
ACK Henry, thanks

On Wed, Aug 26, 2015 at 11:14 PM, Henry Saputra <notifications@github.com>
wrote:

> @lewismc <https://github.com/lewismc> Need to add "This closes #23
> <https://github.com/apache/gora/pull/23>" in the commit comment
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/gora/pull/23#issuecomment-135304353>.
>



-- 
*Lewis*

**Pull request closed without merging**